### PR TITLE
Joindin 735

### PR DESCRIPTION
**Pull request by ericpoe**
This adds type-hinting to satisfy JOINDIN-735

Type-hinting was added for array, PDO, Request, and more. One benefit to this type-hinting is that a bug was uncovered where the Request and PDO params were being passed in the wrong order!

When the parameter types were known, I updated or added docblocks to those methods to show params that are ints, strings, bools, etc.

## Other Changes ##
The abstract `ApiController` and `EmailBaseService` were renamed to emphasize that they are abstract classes, following the pattern of other abstract controller and service classes.

A bug was fixed in an exception message that was attempting to use a non-existent variable.

The `$string` param for one method was renamed since `string` is a PHP7 keyword. Even though we're currently using PHP 5.6, 7 isn't too far away from this project.

### CODE REVIEW

**Comment by heiglandreas** [comment]
Hey Eric! Thanks for the great work! 

But: (you knew there would be one…) Would it be possible to create different merge-requests from this one? One with the renaming of the Classes, one for the type hints and one for the bugfix? That would make the PR much more digestible for review 😉 

And a variable of name ```$string``` should not conflict with the ```string```-keyword in PHP7. But that's only a minor thing ;)

**Comment by ericpoe** [comment]
@heiglandreas The bugfix was so that the tests would pass after adding the typehints, so I don't think that it can be separated. I can certainly separate the renaming of the classes.

File: src/controllers/BaseTalkController.php [hunk 1]
```diff
@@ -81,14 +81,14 @@ public function getUserMapper($db, $request)
...
-     * @param  PDO      $db
      * @param  Request  $request
+     * @param  PDO      $db
      * @param  integer  $talk_id
      * @param  boolean $verbose
      *
      * @throws Exception if the talk is not found
      *
-     * @return TalkModelCollection
+     * @return false|TalkModel
```
Comment (exussum12) [inline src/controllers/BaseTalkController.php]: How can this return false ?

File: tests/models/TestApiMapper.php [hunk 2]
```diff
@@ -26,16 +31,20 @@ public function buildLimit($resultsperpage, $start)
...
 
-    public function inflect($string)
+    /**
+     * @param string $text
+     * @return string
+     */
+    public function inflect($text)
     {
-        return parent::inflect($string); // TODO: Change the autogenerated stub
+        return parent::inflect($text); // TODO: Change the autogenerated stub
```
Comment (exussum12) [inline tests/models/TestApiMapper.php]: Delete the function ?

File: src/views/JsonPView.php [hunk 3]
```diff
@@ -2,8 +2,12 @@
 
 class JsonPView extends JsonView
 {
+    /** @var string */
+    protected $_callback;
+
     public function __construct($callback)
     {
+        parent::__construct();
```
Comment (exussum12) [inline src/views/JsonPView.php]: Nice spot!

**Comment by ericpoe** [comment]
@heiglandreas The abstract class rename has been split out into https://github.com/joindin/joindin-api/pull/425

File: src/controllers/BaseTalkController.php (hunk 1 above)
Comment (ericpoe) [inline src/controllers/BaseTalkController.php]: @exussum12 Good catch! I looked for other examples and found one more. Fixed via bf3d283187

File: tests/models/TestApiMapper.php (hunk 2 above)
Comment (ericpoe) [inline tests/models/TestApiMapper.php]: Consider it done. At least in another PR :) https://github.com/joindin/joindin-api/pull/426

**Comment by ericpoe** [comment]
@heiglandreas the Exception fix is in https://github.com/joindin/joindin-api/pull/427

I rebased to pull out the code changes that are not related to the typehints and miscounted on an interactive rebase. I don't know how to make 1c225d4 not show me as a contributor. Argh!

**Comment by ericpoe** [comment]
Should I kill this branch & PR, create a new branch, cherry pick my changes (minus 1c225d4), and create a new PR?

**Comment by heiglandreas** [comment]
I'm currently reviewing it :wink: So **please not at the moment** 😄 

And I'd leave it but replace it with a forced push…

**Comment by liam-wiltshire** [comment]
ok to test

File: src/models/TalkMapper.php [hunk 4]
```diff
@@ -279,6 +291,12 @@ protected function getTracks($talk_id)
         return $retval;
     }
 
+    /**
+     * @param int $user_id
+     * @param int $resultsperpage
+     * @param int $start
+     * @return bool|TalkModelCollection
```
Comment (heiglandreas) [inline src/models/TalkMapper.php]: Please add a blank line after the ```@param```s

File: src/models/TalkMapper.php [hunk 5]
```diff
@@ -649,6 +667,10 @@ protected function generateInflectedTitle($title, $talk_id)
      * values per event, so you can have the same inflected talk title at two
      * different events, but try to have the same one at the same event and this
      * update will fail.  The calling code catches this and picks a new title
+     *
+     * @param string $inflected_title
+     * @param int $talk_id
+     * @return bool
```
Comment (heiglandreas) [inline src/models/TalkMapper.php]: Please add a blank line after the ```@param```s

File: src/models/TalkMapper.php [hunk 6]
```diff
@@ -332,11 +350,11 @@ public function getTalksBySpeaker($user_id, $resultsperpage, $start)
      * - speakers (an array of names)
      * - type_id (id of the talk's type)
      *
-     * @param $data
-     *
-     * @return int
+     * @param array $data
+     * @return string
```
Comment (heiglandreas) [inline src/models/TalkMapper.php]: Please add a blank line after the ```@param```s

File: src/models/TalkMapper.php [hunk 7]
```diff
@@ -332,11 +350,11 @@ public function getTalksBySpeaker($user_id, $resultsperpage, $start)
...
+     * @throws Exception
```
Comment (heiglandreas) [inline src/models/TalkMapper.php]: Please move the ```@return``` to the last line of the DocBlock

File: src/models/TalkMapper.php [hunk 8]
```diff
@@ -384,11 +402,11 @@ public function createTalk($data)
      * - speakers (an array of names)
      * - type_id (id of the talk's type)
      *
-     * @param $data
-     *
-     * @return void
+     * @param array $data
+     * @param int $talk_id
+     * @throws Exception
```
Comment (heiglandreas) [inline src/models/TalkMapper.php]: Please add a blank line after the ```@param```s

File: src/models/TalkMapper.php [hunk 9]
```diff
@@ -668,6 +690,10 @@ protected function storeInflectedTitle($inflected_title, $talk_id)
         return $result;
     }
 
+    /**
+     * @param int $talk_id
+     * @return array
```
Comment (heiglandreas) [inline src/models/TalkMapper.php]: Please add a blank line after the ```@param```

File: src/models/TalkMapper.php [hunk 10]
```diff
@@ -831,6 +857,11 @@ public function delete($talk_id)
         return true;
     }
 
+    /**
+     * @param int $talk_id
+     * @param string $display_name
+     * @return bool
```
Comment (heiglandreas) [inline src/models/TalkMapper.php]: Please add a blank line after the ```@param```s

File: src/models/TalkMapper.php [hunk 11]
```diff
@@ -855,6 +890,10 @@ public function removeApprovedSpeakerFromTalk($talk_id, $speaker_id)
         $stmt->execute($params);
     }
 
+    /**
+     * @param int $talk_id
+     * @return bool
```
Comment (heiglandreas) [inline src/models/TalkMapper.php]: Please add a blank line after the ```@param```

File: src/models/TalkMapper.php [hunk 12]
```diff
@@ -864,6 +903,13 @@ public function removeAllSpeakersFromTalk($talk_id)
         return $stmt->execute(['talk_id' => $talk_id]);
     }
 
+    /**
+     * @param int $talk_id
+     * @param int $claim_id
+     * @param int $speaker_id
+     * @param string $speaker_name
+     * @return bool
```
Comment (heiglandreas) [inline src/models/TalkMapper.php]: Please add a blank line after the ```@param```s

File: src/models/TalkMapper.php [hunk 13]
```diff
@@ -881,6 +927,11 @@ public function assignTalkToSpeaker($talk_id, $claim_id, $speaker_id, $speaker_n
         );
     }
 
+    /**
+     * @param int $talk_id
+     * @param int|null $link_id
+     * @return array
```
Comment (heiglandreas) [inline src/models/TalkMapper.php]: Please add a blank line after the ```@param```s

File: src/models/TalkMapper.php [hunk 14]
```diff
@@ -912,7 +963,11 @@ public function getTalkMediaLinks($talk_id, $link_id = null)
         return $stmt->fetchAll(PDO::FETCH_ASSOC);
     }
 
-    public function addTalkMediaTypes($talk)
+    /**
+     * @param array $talk
+     * @return array
```
Comment (heiglandreas) [inline src/models/TalkMapper.php]: Please add a blank line after the ```@param```s

File: src/models/TalkMapper.php [hunk 15]
```diff
@@ -924,6 +979,11 @@ public function addTalkMediaTypes($talk)
         return $talk;
     }
 
+    /**
+     * @param int $talk_id
+     * @param int $link_id
+     * @return bool
```
Comment (heiglandreas) [inline src/models/TalkMapper.php]: Please add a blank line after the ```@param```s

File: src/models/TalkMapper.php [hunk 16]
```diff
@@ -960,6 +1020,13 @@ public function removeAllTalkLinks($talk_id)
         return $stmt->execute(['talk_id' => $talk_id]);
     }
 
+    /**
+     * @param int $talk_id
+     * @param int $link_id
+     * @param string$display_name
+     * @param string $url
+     * @return bool
```
Comment (heiglandreas) [inline src/models/TalkMapper.php]: Please add a blank line after the ```@param```s

File: src/models/TalkMapper.php [hunk 17]
```diff
@@ -987,6 +1054,12 @@ public function updateTalkLink($talk_id, $link_id, $display_name, $url)
         return 1 == $stmt->rowCount();
     }
 
+    /**
+     * @param int $talk_id
+     * @param string $display_name
+     * @param string $url
+     * @return string
```
Comment (heiglandreas) [inline src/models/TalkMapper.php]: Please add a blank line after the ```@param```s

File: src/models/TalkMapper.php [hunk 18]
```diff
@@ -1014,8 +1087,12 @@ public function addTalkLink($talk_id, $display_name, $url)
 
     /**
      * Used for adding back the slide link to the request
+     *
+     * @param array $talk
+     * @param array $link
+     * @return array
```
Comment (heiglandreas) [inline src/models/TalkMapper.php]: Please add a blank line after the ```@param```s

File: src/models/TalkModel.php [hunk 19]
```diff
@@ -68,6 +68,10 @@ public function getSubResources()
 
     /**
      * Return this object with client-facing fields and hypermedia, ready for output
+     *
+     * @param Request $request
+     * @param bool $verbose
+     * @return array
```
Comment (heiglandreas) [inline src/models/TalkModel.php]: Please add a blank line after the ```@param```s

File: src/models/TalkModel.php [hunk 20]
```diff
@@ -94,7 +98,7 @@ public function getOutputView(Request $request, $verbose = false)
     /**
      * Get the URL on the website of this talk
      *
-     * @param $website_url string The URL to the main website (e.g. http://joind.in or http://test.joind.in)
+     * @param string $website_url The URL to the main website (e.g. http://joind.in or http://test.joind.in)
      * @return string The link to the talk on the web (e.g. http://web2.dev.joind.in/talk/ed89b)
```
Comment (heiglandreas) [inline src/models/TalkModel.php]: Please add a blank line after the ```@param```s

File: src/models/TalkModelCollection.php [hunk 21]
```diff
@@ -31,8 +38,12 @@ public function __construct(array $data, $total)
      * This creates the expected output structure, converting each resource
      * to it's presentable representation and adding the meta fields for totals
      * and pagination
+     *
+     * @param Request $request
+     * @param bool $verbose
+     * @return array
```
Comment (heiglandreas) [inline src/models/TalkModelCollection.php]: Please add a blank line after the ```@param```s

File: src/models/TalkTypeMapper.php [hunk 22]
```diff
@@ -22,6 +28,12 @@ public function getVerboseFields()
         return $fields;
     }
 
+    /**
+     * @param int $resultsperpage
+     * @param int $start
+     * @param bool $verbose
+     * @return array|bool
```
Comment (heiglandreas) [inline src/models/TalkTypeMapper.php]: Please add a blank line after the ```@param```s

File: src/models/TalkTypeMapper.php [hunk 23]
```diff
@@ -31,15 +43,26 @@ public function getTalkTypeList($resultsperpage, $start = 0, $verbose = false)
         return false;
     }
 
-    public function getTalkTypeById($talkTypeId, $verbose)
+    /**
+     * @param int $talkType_id
+     * @param bool $verbose
+     * @return array|bool
```
Comment (heiglandreas) [inline src/models/TalkTypeMapper.php]: Please add a blank line after the ```@param```s

File: src/models/TalkTypeMapper.php [hunk 24]
```diff
@@ -31,15 +43,26 @@ public function getTalkTypeList($resultsperpage, $start = 0, $verbose = false)
...
             return $this->transformResults($results, $verbose);
         }
         return false;
     }
 
+    /**
+     * @param int $resultsperpage
+     * @param int $start
+     * @param array $params
+     * @return array|bool
```
Comment (heiglandreas) [inline src/models/TalkTypeMapper.php]: Please add a blank line after the ```@param```s

File: src/models/TokenModel.php [hunk 25]
```diff
@@ -39,6 +39,10 @@ public function getVerboseFields()
 
     /**
      * Return this object with client-facing fields and hypermedia, ready for output
+     *
+     * @param Request $request
+     * @param bool $verbose
+     * @return array
```
Comment (heiglandreas) [inline src/models/TokenModel.php]: Please add a blank line after the ```@param```s

File: src/models/TokenModelCollection.php [hunk 26]
```diff
@@ -32,10 +39,13 @@ public function __construct(array $data, $total)
      * to it's presentable representation and adding the meta fields for totals
      * and pagination
      *
-     * @
+     * @param Request $request
+     * @param bool $verbose
+     * @return array
```
Comment (heiglandreas) [inline src/models/TokenModelCollection.php]: Please add a blank line after the ```@param```s

File: src/models/TrackMapper.php [hunk 27]
```diff
@@ -24,6 +30,13 @@ public function getVerboseFields()
         return $fields;
     }
 
+    /**
+     * @param int $event_id
+     * @param int $resultsperpage
+     * @param int $start
+     * @param bool $verbose
+     * @return array|bool
```
Comment (heiglandreas) [inline src/models/TrackMapper.php]: Please add a blank line after the ```@param```s

File: src/models/TrackMapper.php [hunk 28]
```diff
@@ -71,6 +87,11 @@ public function transformResults($results, $verbose)
         return $retval;
     }
 
+    /**
+     * @param int $track_id
+     * @param bool $verbose
+     * @return array|bool
```
Comment (heiglandreas) [inline src/models/TrackMapper.php]: Please add a blank line after the ```@param```s

File: src/models/TrackMapper.php [hunk 29]
```diff
@@ -90,7 +111,13 @@ public function getTrackById($track_id, $verbose = false)
         return false;
     }
 
-    public function createEventTrack($data, $event_id)
+    /**
+     * @param array $data
+     * @param int $event_id
+     * @return string
```
Comment (heiglandreas) [inline src/models/TrackMapper.php]: Please add a blank line after the ```@param```s and move the ```@return``` to the last line of the DocBlock

File: src/models/TwitterRequestTokenMapper.php [hunk 30]
```diff
@@ -2,6 +2,11 @@
 
 class TwitterRequestTokenMapper extends ApiMapper
 {
+    /**
+     * @param string $token
+     * @param string $secret
+     * @return bool|TwitterRequestTokenModelCollection
```
Comment (heiglandreas) [inline src/models/TwitterRequestTokenMapper.php]: Please add a blank line after the ```@param```s

File: src/models/TwitterRequestTokenMapper.php [hunk 31]
```diff
@@ -27,6 +32,10 @@ public function create($token, $secret)
         return false;
     }
 
+    /**
+     * @param string $token
+     * @return bool
```
Comment (heiglandreas) [inline src/models/TwitterRequestTokenMapper.php]: Please add a blank line after the ```@param```s

File: src/models/TwitterRequestTokenModel.php [hunk 32]
```diff
@@ -41,6 +41,10 @@ protected function getVerboseFields()
 
     /**
      * Return this object with client-facing fields and hypermedia, ready for output
+     *
+     * @param Request $request
+     * @param bool $verbose
+     * @return array
```
Comment (heiglandreas) [inline src/models/TwitterRequestTokenModel.php]: Please add a blank line after the ```@param```s

File: src/models/TwitterRequestTokenModelCollection.php [hunk 33]
```diff
@@ -25,8 +34,12 @@ public function __construct(array $data, $total = 0)
      * This creates the expected output structure, converting each resource
      * to it's presentable representation and adding the meta fields for totals
      * and pagination
+     *
+     * @param Request $request
+     * @param bool $verbose
+     * @return array
```
Comment (heiglandreas) [inline src/models/TwitterRequestTokenModelCollection.php]: Please add a blank line after the ```@param```s

File: src/models/UserMapper.php [hunk 34]
```diff
@@ -44,6 +43,11 @@ public function getVerboseFields()
         return $fields;
     }
 
+    /**
+     * @param int $user_id
+     * @param bool $verbose
+     * @return array|bool
```
Comment (heiglandreas) [inline src/models/UserMapper.php]: Please add a blank line after the ```@param```s

File: src/models/UserMapper.php [hunk 35]
```diff
@@ -54,9 +58,13 @@ public function getUserById($user_id, $verbose = false)
         }
 
         return false;
-
     }
 
+    /**
+     * @param string $username
+     * @param bool $verbose
+     * @return array|bool
```
Comment (heiglandreas) [inline src/models/UserMapper.php]: Please add a blank line after the ```@param```s

File: src/models/UserMapper.php [hunk 36]
```diff
@@ -97,6 +108,13 @@ public function getSiteAdminEmails()
         return false;
     }
 
+    /**
+     * @param int $resultsperpage
+     * @param int $start
+     * @param string|null $where
+     * @param string|null $order
+     * @return array|bool
```
Comment (heiglandreas) [inline src/models/UserMapper.php]: Please add a blank line after the ```@param```s

File: src/models/UserMapper.php [hunk 37]
```diff
@@ -134,6 +152,12 @@ protected function getUsers($resultsperpage, $start, $where = null, $order = nul
         return false;
     }
 
+    /**
+     * @param int $resultsperpage
+     * @param int $start
+     * @param bool $verbose
+     * @return array|bool
```
Comment (heiglandreas) [inline src/models/UserMapper.php]: Please add a blank line after the ```@param```s

File: src/models/TwitterRequestTokenModelCollection.php [hunk 38]
```diff
@@ -6,12 +6,21 @@
...
     /**
      * Take arrays of data and create a collection of models; store metadata
+     *
+     * @param array $data
+     * @param int $total
      */
     public function __construct(array $data, $total = 0)
     {
         $this->total = $total;
+        $this->list = [];
```
Comment (heiglandreas) [inline src/models/TwitterRequestTokenModelCollection.php]: I'd move that to a separate PR as it'S not related to the typehints

File: src/models/UserMapper.php [hunk 39]
```diff
@@ -147,6 +171,13 @@ public function getUserList($resultsperpage, $start, $verbose = false)
         return false;
     }
 
+    /**
+     * @param int $event_id
+     * @param int $resultsperpage
+     * @param int $start
+     * @param bool $verbose
+     * @return array|bool
```
Comment (heiglandreas) [inline src/models/UserMapper.php]: Please add a blank line after the ```@param```s

File: src/models/UserMapper.php [hunk 40]
```diff
@@ -217,7 +251,10 @@ public function transformResults($results, $verbose)
         return $retval;
     }
 
-
+    /**
+     * @param int $user_id
+     * @return bool
```
Comment (heiglandreas) [inline src/models/UserMapper.php]: Please add a blank line after the ```@param```s

File: src/models/UserMapper.php [hunk 41]
```diff
@@ -231,7 +268,7 @@ public function isSiteAdmin($user_id)
     /**
      * Check if the user represented by $user_id is trusted
      *
-     * @param $user_id
+     * @param int $user_id
      * @return bool
```
Comment (heiglandreas) [inline src/models/UserMapper.php]: Please add a blank line after the ```@param```s

File: src/models/UserMapper.php [hunk 42]
```diff
@@ -264,6 +301,11 @@ public function setTrustedStatus($trustedStatus, $user_id)
     }
 
 
+    /**
+     * @param int $user
+     * @return bool|string
```
Comment (heiglandreas) [inline src/models/UserMapper.php]: Please add a blank line after the ```@param```s and move the ```@return``` to the last line of the DocBlock

File: src/models/UserMapper.php [hunk 43]
```diff
@@ -307,6 +349,11 @@ public function createUser($user)
         return false;
     }
 
+    /**
+     * @param string $email
+     * @param bool $verbose
+     * @return array|bool
```
Comment (heiglandreas) [inline src/models/UserMapper.php]: Please add a blank line after the ```@param```s

File: src/models/UserMapper.php [hunk 44]
```diff
@@ -361,6 +408,9 @@ public function checkPasswordValidity($password)
     /**
      * Generate and store a token in the email_verification_tokens table for this
      * user, when they use the token to verify, we'll set their status to verified
+     *
+     * @param int $user_id
+     * @return bool|string
```
Comment (heiglandreas) [inline src/models/UserMapper.php]: Please add a blank line after the ```@param```s

File: src/models/UserMapper.php [hunk 45]
```diff
@@ -455,12 +505,19 @@ public function getUserIdFromEmail($email)
      * Used only on test platforms, if the config is enabled
      *
      * Designed to allow creation of verified users for testing purposes
+     *
+     * @param int $user_id
+     * @return bool
```
Comment (heiglandreas) [inline src/models/UserMapper.php]: Please add a blank line after the ```@param```s

File: src/models/UserMapper.php [hunk 46]
```diff
@@ -455,12 +505,19 @@ public function getUserIdFromEmail($email)
...
      */
     public function verifyThisTestUser($user_id)
     {
         return $this->markUserVerified($user_id);
     }
 
+    /**
+     * @param int $user_id
+     * @return bool
```
Comment (heiglandreas) [inline src/models/UserMapper.php]: Please add a blank line after the ```@param```s

File: src/models/UserMapper.php [hunk 47]
```diff
@@ -509,9 +566,10 @@ public function thisUserHasAdminOn($user_id)
      * @param array $user An array of fields to change
      * @param int $userId The user to update
      *
-     * @return bool True if successful
+     * @return bool if successful
+     * @throws Exception
```
Comment (heiglandreas) [inline src/models/UserMapper.php]: Please move the ```@return``` to the last line of the DocBlock

File: src/models/UserMapper.php [hunk 48]
```diff
@@ -607,6 +665,9 @@ public function getEmailByUserId($user_id)
      * Generate and store a token in the password_reset_tokens table for this
      * user, when they come to web2 with this token, we'll let them set a new
      * password
+     *
+     * @param int $user_id
+     * @return bool|string
```
Comment (heiglandreas) [inline src/models/UserMapper.php]: Please add a blank line after the ```@param```s

File: src/models/UserMapper.php [hunk 49]
```diff
@@ -634,8 +695,9 @@ public function generatePasswordResetTokenForUserId($user_id)
      * Check that the token is valid, find out which user this is, save their
      * new password and then delete their other tokens
      *
-     * @param string $token The reset we sent them by email (link goes to web2)
-     * @param string $password The new password they chose
+     * @param string $token
+     * @param string $password
+     * @return bool
```
Comment (heiglandreas) [inline src/models/UserMapper.php]: Please add a blank line after the ```@param```s

File: src/models/UserMapper.php [hunk 50]
```diff
@@ -680,6 +742,10 @@ public function resetPassword($token, $password)
         return false;
     }
 
+    /**
+     * @param int $user_id
+     * @return bool
```
Comment (heiglandreas) [inline src/models/UserMapper.php]: Please add a blank line after the ```@param```

File: src/services/EmailBaseService.php [hunk 51]
```diff
@@ -55,8 +55,12 @@ public function __construct($config, $recipients)
     /**
      * Take the template and the replacements, return markdown
      * with the correct values in it
+     *
+     * @param string $templateName
+     * @param array $replacements
+     * @return string
```
Comment (heiglandreas) [inline src/services/EmailBaseService.php]: Please add a blank line after the ```@param```s

File: src/services/EmailBaseService.php [hunk 52]
```diff
@@ -71,6 +75,9 @@ public function parseEmail($templateName, $replacements)
 
     /**
      * Set the body of the message
+     *
+     * @param string $body
+     * @return $this
```
Comment (heiglandreas) [inline src/services/EmailBaseService.php]: Please add a blank line after the ```@param```s

File: src/services/EmailBaseService.php [hunk 53]
```diff
@@ -83,6 +90,9 @@ protected function setBody($body)
      * Set the HTML body of the message
      *
      * Call setBody first
+     *
+     * @param string $body
+     * @return $this
```
Comment (heiglandreas) [inline src/services/EmailBaseService.php]: Please add a blank line after the ```@param```s

File: src/services/EmailBaseService.php [hunk 54]
```diff
@@ -128,6 +144,9 @@ public function getRecipients()
 
     /**
      * Markdown to HTML
+     *
+     * @param string $markdown
+     * @return string mixed
```
Comment (heiglandreas) [inline src/services/EmailBaseService.php]: Please add a blank line after the ```@param```s

File: src/services/EmailBaseService.php [hunk 55]
```diff
@@ -136,6 +155,10 @@ public function markdownToHtml($markdown)
         return $messageHTML;
     }
 
+    /**
+     * @param string $html
+     * @return string
```
Comment (heiglandreas) [inline src/services/EmailBaseService.php]: Please add a blank line after the ```@param```s

File: src/services/SpamCheckService.php [hunk 56]
```diff
@@ -59,7 +63,12 @@ public function isCommentAcceptable($data, $userIp, $userAgent)
         return false;
     }
 
-    protected function getField($key, $data)
+    /**
+     * @param $key
+     * @param array $data
+     * @return bool|mixed
```
Comment (heiglandreas) [inline src/services/SpamCheckService.php]: Please add a blank line after the ```@param```s

File: src/views/ApiView.php [hunk 57]
```diff
@@ -2,12 +2,26 @@
...
+    public function __construct(array $headers = [], $responseCode = 200, $noRender = false)
+    {
+        $this->headers = $headers;
+        $this->responseCode = $responseCode;
+        $this->noRender = $noRender;
+    }
+
+    /**
+     * @param $content
+     * @return array
```
Comment (heiglandreas) [inline src/views/ApiView.php]: Please add a blank line after the ```@param```s

File: src/views/ApiView.php [hunk 58]
```diff
@@ -24,21 +38,35 @@ protected function addCount($content)
...
+     * @param bool $noRender
+     */
     public function setNoRender($noRender)
     {
         $this->noRender = $noRender;
     }
 
+    /**
+     * @param $content
+     * @return bool
```
Comment (heiglandreas) [inline src/views/ApiView.php]: Please add a blank line after the ```@param```s

File: src/views/HtmlView.php [hunk 59]
```diff
@@ -12,6 +12,10 @@
 class HtmlView extends ApiView
 {
 
+    /**
+     * @param $content
+     * @return bool
```
Comment (heiglandreas) [inline src/views/HtmlView.php]: Please add a blank line after the ```@param```s

File: src/views/JsonPView.php [hunk 60]
```diff
@@ -2,9 +2,13 @@
 
 class JsonPView extends JsonView
 {
+    /** @var string */
+    protected $callback;
+
     public function __construct($callback)
     {
-        $this->_callback = $callback;
+        parent::__construct();
```
Comment (heiglandreas) [inline src/views/JsonPView.php]: That's not necessary at all, as there's no constructor defined in one of the parents.

Additionaly it's not related to the TypeHints, so I'd rather move it to a different PR

File: tests/controllers/TalksControllerTest.php [hunk 61]
```diff
@@ -12,6 +12,17 @@
...
+
+    public function setUp()
+    {
+        $this->config = [
+            'email' => [
+                'from' => 'source@example.com',
+            ],
+            'website_url' => 'http://example.com',
+        ];
+    }
```
Comment (heiglandreas) [inline tests/controllers/TalksControllerTest.php]: These changes are not related to the TypeHints, so I'd move them to a different PR

File: tests/inc/RequestTest.php [hunk 62]
```diff
@@ -425,7 +425,7 @@ public function schemeProvider()
     public function ifIdentificationDoesNotHaveTwoPartsExceptionIsThrown()
     {
         $request = new \Request($this->config, ['HTTPS' => 'on']);
-        $request->identifyUser(null, 'This is a bad header');
+        $request->identifyUser('This is a bad header');
```
Comment (heiglandreas) [inline tests/inc/RequestTest.php]: See the comment on the ```Request```-class.

File: tests/models/TestApiMapper.php [hunk 63]
```diff
@@ -18,6 +18,11 @@ public function getVerboseFields()
         return [];
     }
 
+    /**
+     * @param int $resultsperpage
+     * @param int $start
+     * @return string
```
Comment (heiglandreas) [inline tests/models/TestApiMapper.php]: Please add a blank line after the ```@param```s

File: tests/inc/RequestTest.php [hunk 64]
```diff
@@ -457,7 +457,7 @@ public function ifRequestIsntHTTPSReturnsFalse()
         $config = array_merge($this->config, array('mode' => 'production'));
         $request = new \Request($config, []);
         $request->setScheme('http://');
-        $this->assertFalse($request->identifyUser(null, 'This is a bad header'));
+        $this->assertFalse($request->identifyUser('This is a bad header'));
```
Comment (heiglandreas) [inline tests/inc/RequestTest.php]: See the comment on the ```Request```-class.

File: tests/inc/RequestTest.php [hunk 65]
```diff
@@ -442,7 +442,7 @@ public function ifIdentificationDoesNotHaveTwoPartsExceptionIsThrown()
     public function ifIdentificationHeaderDoesNotStartWithOauthThrowException()
     {
         $request = new \Request($this->config, ['HTTPS' => 'on']);
-        $request->identifyUser(null, 'Auth Me');
+        $request->identifyUser('Auth Me');
```
Comment (heiglandreas) [inline tests/inc/RequestTest.php]: See the comment on the ```Request```-class.

File: tests/inc/RequestTest.php [hunk 66]
```diff
@@ -552,7 +552,7 @@ public function identifyUserWithOauthTokenTypeSetsUserIdForValidHeader()
 
         $request->setOauthModel($mockOauth);
 
-        $request->identifyUser(null, 'oauth authPart');
+        $request->identifyUser('oauth authPart');
```
Comment (heiglandreas) [inline tests/inc/RequestTest.php]: See the comment on the ```Request```-class.

File: tests/inc/RequestTest.php [hunk 67]
```diff
@@ -577,7 +577,7 @@ public function identifyUserWithBearerTokenTypeSetsUserIdForValidHeader()
 
         $request->setOauthModel($mockOauth);
 
-        $request->identifyUser(null, 'Bearer authPart');
+        $request->identifyUser('Bearer authPart');
```
Comment (heiglandreas) [inline tests/inc/RequestTest.php]: See the comment on the ```Request```-class.

File: src/models/ClientModelCollection.php [hunk 68]
```diff
@@ -32,10 +39,13 @@ public function __construct(array $data, $total)
...
      *
-     * @
+     * @param Request $request
+     * @param bool $verbose
+     * @return array
      */
-    public function getOutputView($request, $verbose = false)
+    public function getOutputView(Request $request, $verbose = false)
     {
+        $retval = [];
```
[the remark this answers is not in the document]
Comment (ericpoe) [reply src/models/ClientModelCollection.php]: See #429 :)

File: tests/inc/RequestTest.php (hunk 62 above)
Comment (ericpoe) [reply tests/inc/RequestTest.php]: Which comment would that be? As part of this PR, the params were switched for `identifyUser` so that the param with a default is last in the list.

File: src/models/TalkCommentReportModelCollection.php [hunk 69]
```diff
@@ -30,9 +37,14 @@ public function __construct(array $data, $total)
...
      * and pagination
+     *
+     * @param Request $request
+     * @param bool $verbose
+     * @return array
      */
-    public function getOutputView($request, $verbose = false)
+    public function getOutputView(Request $request, $verbose = false)
     {
+        $retval = [];
```
[the remark this answers is not in the document]
Comment (ericpoe) [reply src/models/TalkCommentReportModelCollection.php]: See #429

File: src/views/JsonPView.php (hunk 60 above)
Comment (ericpoe) [reply src/views/JsonPView.php]: See: #432

File: tests/controllers/TalksControllerTest.php (hunk 61 above)
Comment (ericpoe) [reply tests/controllers/TalksControllerTest.php]: This change is because the `ApiController`'s ctor has a type-hinted config as an array. So this change is essentially mocking the config so the test will pass.

**Comment by liam-wiltshire** [comment]
Wow... my head hurts! ;-)

@heiglandreas as you've mainly been the one reviewing this, are you mostly happy? It seems fine to me, and it doesn't break (which is always the main thing!), but I've not been 'hands on' with this PR :).

**Comment by liam-wiltshire** [comment]
@heiglandreas - did you get a chance to re-vist this? :)

**Comment by heiglandreas** [comment]
I just went through and it looks good for me. There might be some minor details that could be improved but it's a great improvement overall already and so I'd merge it.

**Comment by liam-wiltshire** [comment]
Thanks will do this eve :)

**Comment by ericpoe** [comment]
Thanks for your patience and help with this PR. I'll try to keep my next PRs a little more concise and smaller in scope. :)

**Comment by heiglandreas** [comment]
Well, the scope was pretty small, wasn't it? It was *just typehinting* 😛

**Comment by ericpoe** [comment]
Just think about how much more awesome this PR would be if we added PHP 7.1 typehints & return types!

For another day. ;)
